---
title: 'Metabarcoding - AB332'
output: html_notebook
---

# Metabarcoding Script

```{r}
# Main Wrangling Packages includes ggplot
library(tidyverse)
library(lubridate)
library(glue)
# Helper Packages
library(here)
# Vegan for Indixes and zCompositions for normalizing
library(vegan)
library(zCompositions)
# Plot Helper
library(ggforce)
library(patchwork)
library(ggrepel)
# Misc
library(imputeTS) # Impute missing data

source("functions/functions.R")
```

## Data Informations

The data is 3 year from the Isa time series (2011-2014) processed through USearch with clustering at 99%,
arbitrarily removing any OTU with less than 50 reads in the total dataset and removing OTUs assigned to Streptophyta, Metazoa and Fungi. Taxonomic assignment with PR2 using blast. Only DNA from 25m is included.

## Loading data

```{r}
data_meta <- readr::read_tsv("data/AB332metadata_v3.txt") %>%
    dplyr::select(-date) %>%
    glimpse()
data_taxa <- readr::read_tsv("data/AB332_2021_taxtab.txt") %>%
    glimpse()
data_otu_t <- readr::read_tsv("data/AB332_otutab_reduc3.txt") %>%
    column_to_rownames(var = "OTUNumber") %>%
    # for vegan package
    t()
data_otu_t <- data_otu_t[, -(which(colSums(data_otu_t) == 0))]

data_otu <- readr::read_tsv("data/AB332_otutab_reduc3.txt") %>%
    glimpse() %>%
    # Remove empty OTUs
    filter(!if_all(-OTUNumber, ~ .x == 0)) %>%
    glimpse()
```

## Redefine Seasons

winter = daylength == 0
summer = daylength == 24
spring and autumn the betweens
```{r}
data_meta <- data_meta %>%
    mutate(
        seasons_old = seasons,
        seasons = case_when(
            daylength == 0 ~ "winter",
            daylength == 24 ~ "summer",
            between(month, 0, 6) ~ "spring",
            between(month, 7, 12) ~ "autumn"
        )
    )
```

## Normalize counts

We have different sample reads:

```{r}
vegan::rarecurve(data_otu_t, step = 100, xlab = "Number of reads", ylab = "Richness", col = "blue")
```

For statistical reasons we need to normalize our reads. We could use the standard rarefaction method or the CLR method. For the compositional data plots I tried both variants, the overall pattern is the same but if we look in detail at the MALV groups we loose some with the rarefying function. The compositional bar plots are relative and without read normalizing because you do not test them.

Rarefying:
```{r}
# minimum_reads <- min(rowSums(data_otu_t))
# data_otu_raref <- vegan::rrarefy(data_otu_t, minimum_reads)
# data_otu_raref <- data_otu_raref[, -(which(colSums(data_otu_raref) == 0))]
```

CLR:
```{r}
data_otu_clr_imp <- zCompositions::cmultRepl(t(data_otu_t), output = "p-counts")
data_otu_clr_raw <- compositions::clr(data_otu_clr_imp)
data_otu_clr <- bind_cols(data_otu[, 1], as_tibble(data_otu_clr_raw)) %>% glimpse()
```

## Impute Missing values

There are some missing data, we could drop them or impute them based on a spline function. Both would be fine, I decided to do a imputation, as we would otherwise loose 15 rows.

```{r}
data_meta_imp <- data_meta %>%
    mutate(
        season_nr = as_factor(season_nr),
        watermass = as_factor(watermass)
    ) %>%
    mutate(
        across(where(is.numeric), imputeTS::na_interpolation)
    ) %>%
    glimpse()
```

Missing water masses were (visually) assigned based on the water masses definitions from Skogseth et al. 2020. 

```{r}
watermasses <- tribble(
    ~ext, ~long, ~watermass, ~short, ~temp_low, ~temp_high, ~salinity_low, ~salinity_high,
    "External", "Atlantic Water", 1, "AW", 3, 8, 34.9, 35.5,
    "External", "Arctic Water", 0, "ArW", -2, 0, 34.3, 34.8,
    "Local", "Surface Water", 4, "SW", 1, 8, 33, 34,
    "Local", "Local Water", 5, "LW", -2, 1, 33, 35.5,
    "Local", "Winter-Cooled Water", 6, "WCW", -2, -0.50, 34.4, 35.5,
    "Mixed", "Intermediate Water", 3, "IW", 1.0, 8, 34, 34.7,
    "Mixed", "Transformed Atlantic Water", 2, "TAW", 1, 3, 34.7, 34.9
) %>%
    mutate(
        long = forcats::fct_relevel(long, "Arctic Water", "Atlantic Water", "Surface Water", "Local Water", "Winter-Cooled Water", "Intermediate Water", "Transformed Atlantic Water")
    )
p <- data_meta_imp %>%
    mutate(
        watermass = stringr::str_replace_na(watermass, "Not defined")
    ) %>%
    ggplot(aes(y = temp_25, x = sal_25)) +
    geom_rect(
        data = watermasses %>% filter(ext != "External"),
        mapping = aes(ymin = temp_low, ymax = temp_high, xmin = salinity_low, xmax = salinity_high, fill = long),
        inherit.aes = FALSE,
        alpha = 1
    ) +
    geom_rect(
        data = watermasses %>% filter(ext == "External" | short %in% c("WCW", "TAW")),
        mapping = aes(ymin = temp_low, ymax = temp_high, xmin = salinity_low, xmax = salinity_high, fill = long),
        inherit.aes = FALSE,
        alpha = 1
    ) +
    geom_point(
        color = "black",
        size = 3.5
    ) +
    geom_point(
        aes(color = watermass),
        # color = "black",
        size = 3
    ) +
    ggrepel::geom_label_repel(
        data = data_meta_imp %>% filter(is.na(watermass)),
        aes(label = Sample_Name)
    ) +
    scale_fill_manual(
        values = colorBlindBlack8,
        name = "Water masses\ndefinition"
    ) +
    scale_shape_discrete() +
    scale_y_continuous() +
    scale_x_continuous() +
    # ggrepel::geom_label_repel(
    #    aes(label = paste0(Sample_Name)),
    #    color = "black"
    # ) +
    labs(
        color = "Metadata",
        x = "Salinity [psu]",
        y = "Temperature [°C]"
    )
fSaveImages(p, "water_masses_metadata", w = 8, h = 6)

# Assign Integer for Water masses based on our plot
data_meta_imp <- data_meta_imp %>%
    mutate(
        watermass = case_when(
            Sample_Name == "Isa_120308" ~ 3,
            Sample_Name == "Isa_120419" ~ 2,
            TRUE ~ as.numeric(watermass)
        )
    )

# Plot for showcasing the seasonal pattern of our metabarcoding data based on water masses.
p <- data_meta_imp %>%
    mutate(
        seasons = stringr::str_to_title(seasons)
    ) %>%
    ggplot(aes(y = temp_25, x = sal_25)) +
    geom_rect(
        data = watermasses %>% filter(ext != "External"),
        mapping = aes(ymin = temp_low, ymax = temp_high, xmin = salinity_low, xmax = salinity_high, fill = long),
        inherit.aes = FALSE,
        alpha = 1
    ) +
    geom_rect(
        data = watermasses %>% filter(ext == "External" | short %in% c("WCW", "TAW")),
        mapping = aes(ymin = temp_low, ymax = temp_high, xmin = salinity_low, xmax = salinity_high, fill = long),
        inherit.aes = FALSE,
        alpha = 1
    ) +
    # geom_point(
    #    color = "black",
    #    size = 3.5
    # ) +
    geom_point(
        aes(shape = seasons),
        # color = "black",
        size = 3
    ) +
    scale_fill_manual(
        values = colorBlindBlack8,
        name = "Water masses\ndefinition"
    ) +
    scale_shape_discrete() +
    scale_y_continuous() +
    scale_x_continuous() +
    # ggrepel::geom_label_repel(
    #    aes(label = paste0(Sample_Name)),
    #    color = "black"
    # ) +
    labs(
        color = "Seasons",
        shape = "Seasons",
        x = "Salinity [psu]",
        y = "Temperature [°C]"
    )
fSaveImages(p, "water_masses_seasons_metadata", w = 8, h = 6)
```


## Seasonal Trends of Parameters


```{r}
p <- data_meta_imp %>%
    mutate(
        fake_year = lubridate::as_date(glue("{2010}-{month}-{day}")),
        year = as.factor(year)
    ) %>%
    dplyr::select(year, fake_year, daylength, nitrate, phosphate, silicate, temp_25:virus) %>%
    pivot_longer(-c("year", "fake_year")) %>%
    filter(year != "2011") %>%
    glimpse() %>%
    ggplot(aes(x = fake_year, y = value, color = year, group = year)) +
    geom_line() +
    geom_point() +
    scale_y_continuous(
        labels = scales::label_number_si()
    ) +
    scale_x_date(
        date_breaks = "1 month",
        labels = scales::date_format(format = "%b")
    ) +
    scale_color_grey() +
    labs(
        x = "",
        y = "",
        color = "Years"
    ) +
    facet_wrap(~name, scales = "free_y") +
    theme(
        legend.position = "bottom"
    )

fSaveImages(p, "meta_years", w = 15, h = 10)
```



## Data wrangling

Generating a nice tibble which contains all the information and the normalized counts.

```{r}
# data_df_raref <- as_tibble(data_otu_raref, rownames = NA) %>%
#    rownames_to_column(var = "Sample_Name") %>%
#    pivot_longer(-Sample_Name, names_to = "OTUname") %>%
#    left_join(data_taxa, by = c("OTUname")) %>%
#    left_join(data_meta %>% dplyr::select(-date), by = c("Sample_Name")) %>%
#    glimpse()

data_df_clr <- as_tibble(data_otu_clr, rownames = NA) %>%
    dplyr::rename("OTUname" = "OTUNumber") %>%
    pivot_longer(-OTUname, names_to = "Sample_Name") %>%
    left_join(data_taxa, by = c("OTUname")) %>%
    left_join(data_meta_imp, by = c("Sample_Name")) %>%
    glimpse()

data_df_clr <- data_df_clr %>%
    separate(Sample_Name, into = c("Station", "Date"), sep = "_", remove = FALSE) %>%
    mutate(
        date = lubridate::as_date(Date)
    ) %>%
    glimpse()

data_df <- as_tibble(data_otu, rownames = NA) %>%
    dplyr::rename("OTUname" = "OTUNumber") %>%
    pivot_longer(-OTUname, names_to = "Sample_Name") %>%
    left_join(data_taxa, by = c("OTUname")) %>%
    left_join(data_meta_imp, by = c("Sample_Name")) %>%
    glimpse()

data_df <- data_df %>%
    separate(Sample_Name, into = c("Station", "Date"), sep = "_", remove = FALSE) %>%
    mutate(
        date = lubridate::as_date(Date)
    ) %>%
    glimpse()
```

## Exploration

First plot compositional data with OTU based on Phylum because it is still "good enough" with our metabarcoding data.

```{r}
data_grouped <- data_df %>%
    filter(value != 0) %>%
    group_by(Sample_Name, date, Phylum) %>%
    summarise(
        temp_25 = first(temp_25),
        sal_25 = first(sal_25),
        watermass = first(watermass),
        seasons = first(seasons),
        value = sum(value),
        .groups = "drop"
    ) %>%
    glimpse()
```

Generate relative counts and group everything below 5% relative abundance for better visibility except for our MALVs.

```{r}
data_filtered <- data_grouped %>%
    group_by(Sample_Name) %>%
    mutate(value_sum = sum(value)) %>%
    ungroup() %>%
    mutate(value_p = value / value_sum) %>%
    filter(value != 0) %>%
    mutate(
        label = case_when(
            Phylum == "MALVs" ~ Phylum,
            value_p >= 0.05 ~ ifelse(is.na(Phylum), "Unknown", Phylum),
            TRUE ~ "< 5%"
        ),
        seasons = stringr::str_to_title(seasons)
    ) %>%
    glimpse()
```

Plotting the total composition:

```{r}
# Custom Colors for MALVs
n <- unique(data_filtered$label)
color_values <- scales::hue_pal()(length(n))
names(color_values) <- n
color_values["MALVs"] <- "black"
color_values["< 5%"] <- "gray50"
color_values["Unknown"] <- "gray70"

color_values <- color_values[stringr::str_order(names(color_values))]
# Custom Axis Colors
color_seasons <- c("Winter" = "#0072B2", "Spring" = "#E69F00", "Summer" = "#D55E00", "Autumn" = "#009E73")

data_filtered <- data_filtered %>%
    mutate(
        label = forcats::fct_relevel(label, "< 5%", "Unknown", after = 0L),
        label = forcats::fct_relevel(label, "MALVs", after = Inf),
        label_seasons = case_when(
            seasons != lag(seasons, default = "x") ~ seasons,
            TRUE ~ "--"
        )
    ) %>%
    glimpse()

p_meta <- data_filtered %>%
    mutate(
        year = year(date),
        year = ifelse(year %in% c("2011", "2012"), "2011-2012", year),
    ) %>%
    ggplot(aes(x = as.character(date), y = value_p, fill = label)) +
    geom_col(width = 1) +
    geom_text(
        aes(y = 1, label = watermass, color = seasons),
        check_overlap = TRUE,
        nudge_y = 0.05,
        show.legend = FALSE
    ) +
    # geom_text(
    #    aes(y = 1, label = substring(seasons, 1, 2), color = seasons),
    #    check_overlap = TRUE,
    #    nudge_y = 0.15,
    #    show.legend = FALSE
    # ) +
    # ggplot2::annotate(
    #     "text",
    #     y = 1.1,
    #     x = 0,
    #     label = "WM:"
    # ) +
    # scale_x_date(
    #    date_breaks = "1 month",
    #    date_labels = "%m"
    # ) +
    ggplot2::coord_cartesian(clip = FALSE) +
    scale_x_discrete(
        label = ~ substring(.x, first = 6)
    ) +
    scale_fill_manual(
        values = color_values
    ) +
    scale_color_manual(
        values = color_seasons
    ) +
    scale_y_continuous(
        breaks = scales::pretty_breaks(),
        limits = c(0, 1.1),
        label = scales::label_percent()
    ) +
    labs(
        y = "Relative abundance [%]",
        x = "Month",
        fill = "Phylum"
    ) +
    theme(
        legend.position = "right",
    )

p_meta_full <- p_meta +
    geom_text(
        aes(y = 1, label = substring(label_seasons, 1, 2), color = seasons),
        size = 5,
        check_overlap = TRUE,
        nudge_y = 0.10,
        show.legend = FALSE
    ) +
    facet_wrap(~year, scales = "free_x", ncol = 1) +
    theme(
        axis.text.x = element_text(
            angle = 40,
            hjust = 1,
            size = 8
        )
    )

fSaveImages(p_meta_full, "abundance", h = 18, w = 13)
```

### MALVs Detail
```{r}
data_malvs <- data_filtered %>%
    filter(label == "MALVs") %>%
    add_count(seasons, name = "seasons_n") %>%
    add_count(watermass, name = "watermass_n") %>%
    glimpse()
```

Types of MALVs
```{r}
data_malvs_tax <- data_df %>%
    filter(value != 0 & Phylum == "MALVs") %>%
    group_by(date) %>%
    mutate(value_sum = sum(value)) %>%
    ungroup() %>%
    group_by(date, Tax1) %>%
    summarise(
        watermass = first(watermass),
        temp_25 = first(temp_25),
        sal_25 = first(sal_25),
        seasons = first(seasons) %>% stringr::str_to_title(),
        value_p = sum(value) / first(value_sum),
        Tax1 = stringr::str_replace_all(first(Tax1), "_", " ") %>% stringr::str_remove_all("group")
    ) %>%
    glimpse()

data_malvs_tax <- data_malvs_tax %>%
    mutate(
        year = year(date),
        year = ifelse(year %in% c("2011", "2012"), "2011-2012", as.character(year))
    ) %>%
    glimpse()

p_malvs <- data_malvs_tax %>%
    ggplot(aes(x = as.character(date), y = value_p, fill = Tax1)) +
    geom_col(color = "black") +
    # geom_text(
    #    aes(y = 0.95, label = round(temp_25)),
    #    check_overlap = FALSE,
    #    nudge_y = 0.1,
    #    show.legend = FALSE,
    #    size = 1
    # ) +
    scale_color_manual(
        values = color_seasons
    ) +
    scale_fill_manual(
        values = colorBlindBlack8,
    ) +
    scale_y_continuous(
        breaks = scales::pretty_breaks(),
        labels = scales::label_percent()
    ) +
    scale_x_discrete(
        label = ~ substring(.x, first = 6)
    ) +
    labs(
        y = "Relative Abundance MALV [%]",
        x = "Dates",
        fill = "Taxa MALVs"
    )

p_malvs_long <- p_malvs +
    geom_text(
        data = data_malvs_tax %>%
            group_by(date) %>%
            filter(row_number() == 1) %>%
            ungroup() %>%
            mutate(
                label_seasons = case_when(
                    seasons != lag(seasons, default = "x") ~ seasons,
                    TRUE ~ "--"
                )
            ),
        aes(y = 1, label = substring(label_seasons, 1, 2), color = seasons),
        check_overlap = TRUE,
        nudge_y = 0.05,
        size = 2,
        show.legend = FALSE
    ) +
    facet_wrap(~year, scales = "free_x", nrow = 1) +
    theme(
        strip.background = element_blank(),
        strip.text = element_blank(),
        axis.text.x = element_text(angle = 40, hjust = 1, size = 5),
        panel.spacing.x = unit(0, "lines")
    )

p_meta_long <- p_meta +
    geom_text(
        aes(y = 1, label = substring(label_seasons, 1, 2), color = seasons),
        size = 2,
        check_overlap = TRUE,
        nudge_y = 0.10,
        show.legend = FALSE
    ) +
    xlab("") +
    facet_wrap(~year, scales = "free_x", nrow = 1) +
    theme(
        panel.spacing.x = unit(0, "lines"),
        axis.text.x = element_text(angle = 40, hjust = 1, size = 5),
    )

p_malvs_high <- p_malvs +
    geom_text(
        data = data_malvs_tax %>%
            group_by(date) %>%
            filter(row_number() == 1) %>%
            ungroup() %>%
            mutate(
                label_seasons = case_when(
                    seasons != lag(seasons, default = "x") ~ seasons,
                    TRUE ~ "--"
                )
            ),
        aes(y = 1, label = substring(label_seasons, 1, 2), color = seasons),
        check_overlap = TRUE,
        nudge_y = 0.12,
        size = 5,
        show.legend = FALSE
    ) +
    geom_text(
        aes(y = 1, label = watermass, color = seasons),
        check_overlap = TRUE,
        nudge_y = 0.05,
        show.legend = FALSE
    ) +
    facet_wrap(~year, scales = "free_x", ncol = 1) +
    theme(
        axis.text.x = element_text(
            angle = 40,
            hjust = 1,
            size = 8
        )
    )


fSaveImages(p_meta_long / p_malvs_long + patchwork::plot_annotation(tag_levels = "A"), "malvs_abundance", h = 10, w = 18)
fSaveImages(p_malvs_high, "malvs_taxa", h = 18, w = 13)
```


### Season and Temperature / Salinity

```{r}
p1 <- data_malvs %>%
    dplyr::mutate(
        seasons = forcats::fct_relevel(seasons, "Winter", "Spring", "Summer", "Autumn")
    ) %>%
    ggplot(aes(x = seasons, y = value_p, label = glue("n = {seasons_n}"))) +
    ggforce::geom_sina(
        show.legend = FALSE,
        color = colorBlindBlack8[1],
        alpha = 0.5,
        scale = "count"
    ) +
    geom_boxplot(aes(color = seasons), outlier.alpha = 0, show.legend = FALSE, alpha = 0) +
    stat_summary(
        fun = mean, geom = "point", show.legend = FALSE, color = colorBlindBlack8[[8]], shape = 18, size = 4
    ) +
    geom_text(aes(y = 0.45), check_overlap = TRUE) +
    labs(
        y = "Relative abundance [%]",
        x = "Seasons",
    ) +
    scale_color_manual(
        values = color_seasons
    ) +
    scale_y_continuous(
        breaks = scales::pretty_breaks(),
        label = scales::label_percent(accuracy = 1)
    ) +
    theme(
        panel.grid.major.y = element_line()
    )

p2 <- data_malvs %>%
    drop_na(watermass) %>%
    ggplot(aes(x = as.character(watermass), y = value_p, label = glue("n = {watermass_n}"))) +
    ggforce::geom_sina(
        aes(color = seasons),
        show.legend = FALSE,
        alpha = 0.5,
        scale = "count"
    ) +
    geom_boxplot(outlier.alpha = 0, show.legend = F, alpha = 0) +
    stat_summary(
        fun = mean, geom = "point", show.legend = F, color = colorBlindBlack8[[8]], shape = 18, size = 4
    ) +
    geom_text(aes(y = 0.45), check_overlap = TRUE) +
    scale_color_manual(
        values = color_seasons
    ) +
    labs(
        x = "Water masses",
        y = ""
    ) +
    scale_y_continuous(
        breaks = scales::pretty_breaks(),
        label = scales::label_percent(accuracy = 1)
    ) +
    theme(
        panel.grid.major.y = element_line()
    )

p3 <- data_malvs %>%
    arrange(desc(value_p)) %>%
    ggplot(aes(x = temp_25, y = sal_25, color = value_p * 100, size = value_p * 100)) +
    geom_point() +
    scale_y_continuous(
        breaks = scales::pretty_breaks(6)
    ) +
    scale_x_continuous(
        breaks = scales::pretty_breaks(6),
    ) +
    scale_color_viridis_c(limits = c(0, 40), breaks = seq(0, 40, by = 5)) +
    scale_size_continuous(limits = c(0, 40), breaks = seq(0, 40, by = 5)) +
    labs(
        y = "Practical Salinity",
        x = "Temperature [°C]",
        color = "Relative\nabundance [%]",
        size = "Relative\nabundance [%]"
    ) +
    guides(color = guide_legend(), size = guide_legend()) +
    theme(
        panel.grid.major = element_line()
    )


p4 <- data_malvs %>%
    arrange(desc(value_p)) %>%
    ggplot(aes(x = temp_25, y = sal_25, color = seasons, size = value_p * 100)) +
    geom_point() +
    scale_y_continuous(
        breaks = scales::pretty_breaks(6)
    ) +
    scale_x_continuous(
        breaks = scales::pretty_breaks(6),
    ) +
    scale_color_manual(values = color_seasons) +
    scale_size_continuous(limits = c(0, 40), breaks = seq(0, 40, by = 10)) +
    labs(
        y = "Practical Salinity",
        x = "Temperature [°C]",
        color = "Seasons",
        size = "Relative\nabundance [%]"
    ) +
    guides(color = guide_legend(override.aes = list(size = 5))) +
    theme(
        panel.grid.major = element_line()
    )

fSaveImages(p1 + p2 + patchwork::plot_annotation(tag_levels = "A"), "malv_season_watermasses", w = 10)
fSaveImages(p1 + p3 + patchwork::plot_annotation(tag_levels = "A"), "malv_season_saltemp", w = 10)
fSaveImages(p1 + p4 + patchwork::plot_annotation(tag_levels = "A"), "malv_season_saltemp_seasons", w = 10)
```

mMDS on MALVs
```{r}
data_otu_clr_malv <- data_otu_clr %>%
    left_join(data_taxa, by = c("OTUNumber" = "OTUname")) %>%
    filter(Phylum == "MALVs") %>%
    dplyr::select(OTUNumber:Isa_141024) %>%
    glimpse()

m_clr <- vegan::metaMDS(t(data_otu_clr_malv[, -1]), k = 2, trymax = 100, trace = FALSE, autotransform = FALSE, distance = "euclidean")
m_clr_scores <- as_tibble(scores(m_clr), rownames = "Sample_Name") %>%
    left_join(data_meta_imp, by = c("Sample_Name")) %>%
    mutate(seasons = stringr::str_to_title(seasons))
data_meta_scaled <- data_meta_imp %>%
    dplyr::select(daylength:N_P, temp_25:virus) %>%
    mutate(
        across(-c("daylength"), ~ as.numeric(scale(., center = TRUE, scale = TRUE)))
    ) %>%
    glimpse()

m_clr_envfit <- envfit(m_clr, data_meta_scaled)
# plot(m_clr, type = "t", display = "sites")
# plot(m_clr_envfit)

# Create Contour
m_clr_ordi_fluor <- ordisurf(m_clr ~ F_25, data_meta, plot = FALSE)
ordi_grid <- m_clr_ordi_fluor$grid
ordi_data <- expand.grid(x = ordi_grid$x, y = ordi_grid$y)
ordi_data$z <- as.vector(ordi_grid$z)
df_surf <- as_tibble(ordi_data) %>% drop_na()

env_names <- tibble(
    env = c("daylength", "nitrate", "phosphate", "silicate", "N_P", "temp_25", "sal_25", "F_25", "chla_GFF", "chla_10", "chl_small"),
    long = c("DayLength", "Nitrate", "Phosphate", "Silicate", "N/P", "Temperature", "Salinity", "Fluoresence", "GFF", "Chlorophyll10", "ChlorophyllSmall")
)

m_clr_envfit_vectors <- as_tibble(m_clr_envfit$vectors$arrows, rownames = "env") %>%
    add_column(p_value = m_clr_envfit$vectors$pvals) %>%
    left_join(env_names) %>%
    mutate(
        p_rank = rank(desc(p_value), ties.method = "min") * 8,
        label_pos_nmds1 = NMDS1 * 1.15,
        label_pos_nmds2 = NMDS2 * 1.15,
    ) %>%
    glimpse() %>%
    filter(p_value <= 0.05 & !(env %in% c("F_25")))

glimpse(m_clr_scores)
watermasses



p <- m_clr_scores %>%
    left_join(watermasses, by = ("watermass" <- "watermass")) %>%
    ggplot(aes(x = NMDS1, y = NMDS2, color = as.factor(short))) +
    geom_contour_filled(
        data = df_surf, aes(x = x, y = y, z = z, fill = as.numeric(..level..)),
        bins = 20,
        inherit.aes = FALSE,
        alpha = 0.3
    ) +
    geom_point() +
    geom_segment(
        data = m_clr_envfit_vectors,
        aes(y = 0, x = 0, xend = NMDS1 * p_rank, yend = NMDS2 * p_rank),
        inherit.aes = FALSE,
        color = "gray70",
        # alpha = 0.7,
        arrow = arrow(length = unit(2, "mm")),
        lwd = 0.5,
    ) +
    scale_fill_viridis_c() +
    # scale_color_manual(values = color_seasons) +
    scale_color_manual(values = colorBlindBlack8) +
    labs(color = "Water\nmasses", fill = "Fluorescence") +
    guides(color = guide_legend(override.aes = list(size = 3))) +
    ggrepel::geom_text_repel(
        data = m_clr_envfit_vectors,
        aes(label = long, x = label_pos_nmds1 * p_rank, y = label_pos_nmds2 * p_rank),
        min.segment.length = 100,
        size = 3,
        # force = 5,
        force_pull      = 20,
        inherit.aes = FALSE,
    ) +
    coord_fixed(ratio = 1) +
    theme(legend.position = "bottom")


p_seasons <- m_clr_scores %>%
    ggplot(aes(x = NMDS1, y = NMDS2, shape = seasons)) +
    geom_contour_filled(
        data = df_surf, aes(x = x, y = y, z = z, fill = as.numeric(..level..)),
        bins = 20,
        inherit.aes = FALSE,
        alpha = 0.3,
        show.legend = FALSE
    ) +
    geom_point() +
    geom_segment(
        data = m_clr_envfit_vectors,
        aes(y = 0, x = 0, xend = NMDS1 * p_rank, yend = NMDS2 * p_rank),
        inherit.aes = FALSE,
        color = "gray70",
        # alpha = 0.7,
        arrow = arrow(length = unit(2, "mm")),
        lwd = 0.5,
    ) +
    scale_fill_viridis_c() +
    scale_color_manual(values = color_seasons) +
    # scale_color_manual(values = colorBlindBlack8) +
    labs(shape = "Seasons", color = "Seasons", fill = "Fluorescence") +
    guides(shape = guide_legend(override.aes = list(size = 5))) +
    ggrepel::geom_text_repel(
        data = m_clr_envfit_vectors,
        aes(label = long, x = label_pos_nmds1 * p_rank, y = label_pos_nmds2 * p_rank),
        min.segment.length = 100,
        size = 3,
        # force = 5,
        force_pull      = 20,
        inherit.aes = FALSE,
    ) +
    coord_fixed(ratio = 1) +
    theme(legend.position = "bottom")

fSaveImages(p + p_seasons + patchwork::plot_annotation(tag_levels = "A"), "nmds_watermass", w = 12, h = 8)
```


#### Statistical Models

Linear Model on MALVs Community.

```{r}
# library(nlme)
# r_m <- data_malvs %>%
#    mutate(
#        year = lubridate::year(date)
#    ) %>%
#    nlme::lme(
#        fixed = value_p ~ seasons + temp_25 + sal_25,
#        random = ~ 1 | year,
#        data = .,
#    )
# summary(r_m)

library(ggh4x)

ggplot(data_malvs, aes(value_p, fill = seasons)) +
    geom_histogram()

ggplot(data_malvs, aes(sample = value_p)) +
    geom_qq_line() +
    geom_qq() +
    facet_wrap(~seasons)

r_m_a <- lm(value_p ~ seasons, data = data_malvs)
r_m_b <- lm(value_p ~ seasons + temp_25, data = data_malvs)
r_m_c <- lm(value_p ~ seasons + temp_25 + sal_25, data = data_malvs)
anova(r_m_b, r_m_c)

# df_future = data_malvs %>%
#    mutate(
#        year = year(date)
#    ) %>%
#    filter(year == "2012") %>%
#    mutate(
#        temp_25 = temp_25 + 2,
#    )

sec_temp <- ggh4x::help_secondary(
    data_malvs,
    primary = value_p,
    secondary = temp_25,
    name = "Temperature [°C]",
    method = "range",
    breaks = c(-1, 0, 1, 2, 3, 4, 5)
)

p <- data_malvs %>%
    add_column(fitted = fitted(r_m)) %>%
    ggplot(aes(x = as.character(date))) +
    geom_path(aes(y = sec_temp$proj(temp_25), color = "Temperature", group = 1), linetype = "dashed") +
    geom_path(aes(x = as.character(date), y = fitted, color = "Linear Model", group = 1), size = 3) +
    geom_path(aes(x = as.character(date), y = value_p, color = "Original", group = 1), size = 3) +
    scale_y_continuous(
        breaks = scales::pretty_breaks(),
        labels = scales::label_percent(),
        sec.axis = sec_temp
    ) +
    scale_color_manual(
        values = c("Temperature" = "#D55E00", "Linear Model" = "#E69F00", "Original" = "#009E73")
    ) +
    labs(
        x = "Date",
        y = "MALV relative abundance [%]",
        color = ""
    ) +
    theme(
        legend.position = "bottom",
        axis.text.x = element_text(angle = 40, hjust = 1, size = 5)
    )

fSaveImages(p, "LM", h = 6, w = 10)
```



